url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/dense/misc/RrefGaussJordanRowPivot.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) 2009-2014, Peter Abeles. All Rights Reserved.
+ *
+ * This file is part of Efficient Java Matrix Library (EJML).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ejml.alg.dense.misc;
+
+import org.ejml.data.DenseMatrix64F;
+import org.ejml.interfaces.linsol.ReducedRowEchelonForm;
+
+/**
+ * Reduction to RREF using Gauss-Jordan elimination with row (partial) pivots.
+ *
+ * @author Peter Abeles
+ */
+public class RrefGaussJordanRowPivot implements ReducedRowEchelonForm<DenseMatrix64F> {
+
+    // tolerance for singular matrix
+    double tol;
+
+    @Override
+    public void setTolerance(double tol) {
+        this.tol = tol;
+    }
+
+    @Override
+    public void reduce( DenseMatrix64F A , int coefficientColumns) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.